File: 알고리즘 스터디 문제들/Part 7 - 백트래킹/문제/스타트와 링크.cpp
#include <iostream>
#include <array>
using namespace std;

int N;
int min_diff = INT32_MAX;

array<array<int, 20>, 20> stats;
array<bool, 20> choice = {};

void backtrack(int pre, int cnt)
{
    if (cnt == N / 2)
    {
        int start = 0;
        int link = 0;
        for (size_t i = 0; i < N; i++)
        {
            for (size_t j = 0; j < N; j++)
            {
                if (choice[i] && choice[j])
                    start += stats[i][j];
                else if (!choice[i] && !choice[j])
                    link += stats[i][j];
            }
        }

        min_diff = min(min_diff, abs(start - link));
        return;
    }

    for (size_t i = pre + 1; i < N; i++)
    {
        choice[i] = true;
        backtrack(i, cnt + 1);
        choice[i] = false;
    }
}

int main()
{
    cin >> N;
    for (size_t i = 0; i < N; i++)
    {
        for (size_t j = 0; j < N; j++)
            cin >> stats[i][j];
    }

    backtrack(0, 0);
    cout << min_diff;
}

File: 알고리즘 스터디 문제들/Part 7 - 백트래킹/문제/스타트와 링크.py
min_diff = 10000


def backtrack(cnt, stats, used, pre, N):
    global min_diff
    if (cnt == N // 2):
        start = 0
        link = 0
        for i in range(N):
            for j in range(N):
                if (used[i] and used[j]):
                    start += stats[i][j]
                elif ((not used[i]) and (not used[j])):
                    link += stats[i][j]
        min_diff = min(min_diff, abs(start - link))
        return

    for i in range(pre + 1, N):
        used[i] = True
        backtrack(cnt + 1, stats, used, i, N)
        used[i] = False


N = int(input())
stats = [0 for i in range(N)]
used = [False for i in range(N)]
for i in range(N):
    stats[i] = list(map(int, input().split()))

backtrack(0, stats, used, 0, N)
print(min_diff)
